### PR TITLE
ext_proc: refactoring onData() to make it modularized

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
ext_proc: refactoring onData() to make it modularized based on the comments:
https://github.com/envoyproxy/envoy/pull/34942/files#r1829466794

Fixes: issue: #36970

### CODE REVIEW

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 1]
```diff
@@ -553,6 +553,124 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_st
...
+      return FilterDataStatus::Continue;
+    case StreamOpenState::Ok:
+      // Fall through
+      break;
+    }
+
+    // The body has been buffered and we need to send the buffer
+    ENVOY_LOG(debug, "Sending request body message");
+    state.addBufferedData(data);
+    auto req = setupBodyChunk(state, *state.bufferedData(), end_stream);
```
Comment (jmarantz) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:572]: explitic type

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 2]
```diff
@@ -553,6 +553,124 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_st
...
+  // the ability to modify each chunk, in order. Doing this any other way would have required
+  // substantial changes to the filter manager. See
+  // https://github.com/envoyproxy/envoy/issues/16760 for a discussion.
+  switch (openStream()) {
+  case StreamOpenState::Error:
+    return FilterDataStatus::StopIterationNoBuffer;
+  case StreamOpenState::IgnoreError:
+    return FilterDataStatus::Continue;
+  case StreamOpenState::Ok:
+    // Fall through
```
Comment (jmarantz) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:608]: ABSL_FALLTHROUGH

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 3]
```diff
@@ -553,6 +553,124 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_st
...
+    break;
+  }
+
+  // Need to first enqueue the data into the chunk queue before sending.
+  auto req = setupBodyChunk(state, data, end_stream);
```
Comment (jmarantz) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:613]: explicit type

**Review by jmarantz — COMMENTED** [review]
This is a huge improvement. We should get this in first. Then the larger PR will be a lot easier to understand.

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 2 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:608]: I don't think we want to add ABSL_FALLTHROUGH_INTENDED here.
Do you mean add it at line 721?

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 1 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: done

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 3 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: done

**Comment by yanjunxiang-google** [comment]
> this should be easy to merge once the small changes requested are done.
> 
> /wait

Yeah, we should merge this ASAP so the MxN PR won't be blocked on this.

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 2 above)
Comment (jmarantz) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:608]: oh, sorry. The "fall through" comment confused me. We don't need one here.

I am wondering what "fall through" means before a 'break'.

**Review by jmarantz — APPROVED** [review]

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 2 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:608]: Yeah, it's obvious. But, let's just leave it there.
Thanks for approval.  Could you please merge it so I can perform an upstream merge for PR: https://github.com/envoyproxy/envoy/pull/34942?

**Pull request merged**